gateway-js/src/index.ts

```diff
@@ -580,20 +581,21 @@ export class ApolloGateway implements GraphQLService {
 
     this.compositionId = result.id;
     this.supergraphSdl = supergraphSdl;
+    this.supergraphSchema = supergraphSchema.toGraphQLJSSchema();
 
     if (!supergraphSdl) {
       this.logger.error(
         "A valid schema couldn't be composed. Falling back to previous schema.",
       );
     } else {
-      this.updateWithSchemaAndNotify(schema, supergraphSdl);
+      this.updateWithSchemaAndNotify(supergraphSchema, supergraphSdl);
 
       if (this.experimental_didUpdateSupergraph) {
         this.experimental_didUpdateSupergraph(
           {
             compositionId: result.id,
             supergraphSdl,
-            schema: schema.toGraphQLJSSchema(),
+            schema: this.schema!,
```

Comment:
**question** Should this be `this.supergraphSchema`? Previously it was "the thing returned by `createSchemaFromSupergraphSdl` with `toGraphQLJSSchema` called", and now it's gonna be "the thing returned by `createSchemaFromSupergraphSdl`"...right?

Reply:
Yeah good question as well.

This still pass a `GraphQLSchema` object, but it essentially switch from passing the supergraph to passing the API schema. The reason I did that is that the prior code stroke me as inconsistent. That is, `experimental_didUpdateSupergraph` takes 2 arguments of the form `new stuff, old stuff`, but the pre-existing code was passing the new supergraph schema for the new stuff but the API schema for the old stuff, so one of them was wrong in my book.

Now, I didn't find any place that would document which one it should definitively be, so it's everyone guess what users might actually expect. But from the POV of what is publicly exposed, `ApolloGateway.schema` is the API schema, so it kind of made a bit more sense to me that `CompositionInfo.schema` also be the API schema (add to that the fact that we also separately pass `supergraphSdl`, so if you care about the supergraph schema, you can just parse that).

But anyway, it is a small change of behaviour and I should have mentioned it, sorry. If we've uncomfortable doing it here, that's ok, I can switch back to the old (imo wrong but...) behaviour.

Fwiw, I suspect this doesn't matter too much. I think few users probably use `experimental_didUpdateSupergraph`, and those that do likely only care about the API schema (I mean, the supergraph schema is really mostly just the API schema + additional join spec directives that are not well documented, so I'm not sure how users could rely on the latter). It's also called "experimental" :).

Anyway, that was my reasoning here and I definitively should have been more explicit about it upfront. And I don't mind "reverting" the change if you prefer?

Reply:
as always, i appreciate the detailed response.

the previous code kinda made sense to me that we'd be getting both the supergraph SDL text as well as the object that represents the parsed version of said text so someone wouldn't need to re-parse if you care about the supergraph. though i suspect you're right that what users really want is the API schema.

another point toward keeping this change: i think the prevailing sentiment is that we'd deprecate this function and point users toward `onSchemaLoadOrUpdateListeners` (which btw does exactly what your change does e.g. provides the API schema as an object and the supergraph SDL as text).

**TL;DR** Let's change it!